<a href="https://colab.research.google.com/github/Seledalia/TPs-genomica-/blob/main/tp1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Trabajo Práctico:  
# Introducción a análisis genómicos en línea de comando



El análisis genómico moderno depende en gran medida del uso de la línea de comando, que permite ejecutar programas de forma eficiente y reproducible sobre grandes volúmenes de datos.   

Estas herramientas suelen utilizarse en sistemas operativos basados en **Linux**, una familia de software libre que provee estabilidad y flexibilidad. **Ubuntu** es una de sus distribuciones más populares, ampliamente usada en ciencia por su facilidad de instalación y comunidad activa.

En paralelo, plataformas en la nube como **Google Colab** ofrecen un entorno accesible para ejecutar comandos y scripts en diferentes lenguajes de programación (por ejemplo, R o python), integrando la potencia de Linux, sin necesidad de instalaciones locales, permitiendo una aproximación inicial más simple a los análisis genómicos.



## Objetivos  

*   Familiarizarse con el uso de aplicaciones de línea de comando.
*   Aprender comandos básicos de Unix.
*   Entender conceptos básicos sobre rutas y directorios.
*   Familiarizarse con estructuras de datos que serán utilizadas a lo largo de la cursada.
*   Entender las ventajas del uso de entornos virtuales para la ejecución de programas.

## 0. La terminal y Colab

La **terminal** es la interfaz de texto donde el usuario ingresa instrucciones directamente al sistema, en contraste con los entornos gráficos que se basan en el uso del cursor para la ejecución de programas. Existen diferentes tipos de terminales en sistemas tipo Unix y en Windows. Cada una tiene sus características y sintaxis (lenguaje).

En linux, la terminal utiliza el lenguaje Bash. Este lenguaje utilizaremos principalmente para la ejecución de programas durante el curso, a través de la interfaz de colab.

A diferencia de la terminal de Ubuntu,
este documento es entorno interactivo denominado <strong>cuaderno de Colab</strong> que permite escribir texto formateado y ejecutar código.

---
## 1. Ejercicio 1

El objetivo de este primer ejercicio es entender la estructura de directorios de la nube virtual de este colab, la cual es una computadora con el sistema operativo Ubuntu.  

Además, aprenderemos algunos comandos básicos.  

**Nota** ❗
Para ejecutar comandos de bash dentro de colab, debemos utilizar el **signo !** antes del comando, para especificar el lenguaje en que deberá ser ejecutado el código. En algunos casos, el signo ! debe ser reemplazado por **%** para poder ejecutar ciertos programas.  

Tener en cuenta que esto no aplica para una terminal nativa de Unix (Ubuntu).

In [ ]:
# Dentro de la programación el símbolo # (hashtag) representa comentarios
# Todo el texto en estas líneas no son leídos como código!

In [10]:
# en qué directorio me encuentro?
!pwd

/content


/content es el **camino absoluto** al directorio actual. Es decir, la posición de la carpeta en la que nos encontramos, en relación al sistema de discos del computador, denominado **base** y denotado por /.


In [2]:
# qué hay dentro de este directorio?
!ls

sample_data


El comando **ls** significa "list", es decir, listar.  

En general los comandos tienen parámetros que nos permiten controlar la ejecución de programas.

Por ejemplo,  al agregar el parámetro **"-l"** al comando ls, estamos pidiéndole más información sobre archivos y carpetas que se encuentran en este directorio.

In [3]:
# qué es sample data?
!ls -l

total 4
drwxr-xr-x 1 root root 4096 Mar 30 13:29 sample_data


In [4]:

!echo 'Primera letra del resultado nos está informando que se trata de un directorio'

Primera letra del resultado nos está informando que se trata de un directorio


In [5]:
# Qué hay dentro de "sample_data" ?

!ls sample_data

anscombe.json		      mnist_test.csv
california_housing_test.csv   mnist_train_small.csv
california_housing_train.csv  README.md


En este caso, "sample_data" es utilizado como argumento.  En el contexto de Linux o cualquier terminal de comandos, un **argumento** es un valor que se le pasa a un comando para que modifique su comportamiento o le indique sobre qué elemento actuar.

In [8]:
# qué tipo de archivo es README.md?
!ls -l sample_data/README.md


-rwxr-xr-x 1 root root 962 Jan  1  2000 sample_data/README.md


In [9]:
!ls -l sample_data/README.md
!echo 'El primer caracter - significa que se trata de un archivo'

-rwxr-xr-x 1 root root 962 Jan  1  2000 sample_data/README.md
El primer caracter - significa que se trata de un archivo


Nótese que para poder listar el archivo README.md fue necesario indicar el "camino" completo hacia ese archivo.  
Es decir, sample_data/README.md es el **camino relativo** a la posicion actual.

In [11]:
# Ingresar al directorio sample_data
%cd sample_data

/content/sample_data


A continuación complete el código para entender la localización actual.

In [12]:
# En qué directorio estoy? cual es su camino absoluto?
!pwd






/content/sample_data


Estoy en el directorio de sample_data, dentro de content


In [13]:
# Qué archivos se encuentran? Qué son?
!ls -l

total 55504
-rwxr-xr-x 1 root root     1697 Jan  1  2000 anscombe.json
-rw-r--r-- 1 root root   301141 Mar 30 13:29 california_housing_test.csv
-rw-r--r-- 1 root root  1706430 Mar 30 13:29 california_housing_train.csv
-rw-r--r-- 1 root root 18289443 Mar 30 13:29 mnist_test.csv
-rw-r--r-- 1 root root 36523880 Mar 30 13:29 mnist_train_small.csv
-rwxr-xr-x 1 root root      962 Jan  1  2000 README.md


Son archivos de texto plano en su mayoria, hay un archivo .json

In [19]:
# Volver al directorio anterior.
%cd ..



/content


## 2. Ejercicio 2
Interacción con archivos

crear carpetas
archivos
leer archivos


In [24]:
# Crear un archivo en blanco
!touch datos.csv
!ls

datos.csv  sample_data


In [26]:
# Eliminar el archivo
!rm datos.csv
!ls

rm: cannot remove 'datos.csv': No such file or directory
sample_data


In [28]:
# Crear un directorio
!mkdir resultados


# Ver si fue creado
!ls




mkdir: cannot create directory ‘resultados’: File exists
resultados  sample_data


In [30]:
# Eliminar un directorio
!rmdir resultados

# Ver si fue eliminado
!ls


rmdir: failed to remove 'resultados': No such file or directory
sample_data


In [31]:
# Imprimir el contenido de un archivo
!cat sample_data/README.md

This directory includes a few sample datasets to get you started.

*   `california_housing_data*.csv` is California housing data from the 1990 US
    Census; more information is available at:
    https://docs.google.com/document/d/e/2PACX-1vRhYtsvc5eOR2FWNCwaBiKL6suIOrxJig8LcSBbmCbyYsayia_DvPOOBlXZ4CAlQ5nlDD8kTaIDRwrN/pub

*   `mnist_*.csv` is a small sample of the
    [MNIST database](https://en.wikipedia.org/wiki/MNIST_database), which is
    described at: http://yann.lecun.com/exdb/mnist/

*   `anscombe.json` contains a copy of
    [Anscombe's quartet](https://en.wikipedia.org/wiki/Anscombe%27s_quartet); it
    was originally described in

    Anscombe, F. J. (1973). 'Graphs in Statistical Analysis'. American
    Statistician. 27 (1): 17-21. JSTOR 2682899.

    and our copy was prepared by the
    [vega_datasets library](https://github.com/altair-viz/vega_datasets/blob/4f67bdaad10f45e3549984e17e1b3088c731503d/vega_datasets/_data/anscombe.json).


In [36]:
# Imprimir 50 primeras lineas de un archivo
head -50 sample_data/california_housing_test.csv


SyntaxError: invalid syntax (262877499.py, line 2)

**Cuál fue el problema?** 🤔 corregir y volver a correr.

Que al inicio de la linea de codigo no se puso la salida a la terminal (!)

In [37]:
!head -50 sample_data/california_housing_test.csv

"longitude","latitude","housing_median_age","total_rooms","total_bedrooms","population","households","median_income","median_house_value"
-122.050000,37.370000,27.000000,3885.000000,661.000000,1537.000000,606.000000,6.608500,344700.000000
-118.300000,34.260000,43.000000,1510.000000,310.000000,809.000000,277.000000,3.599000,176500.000000
-117.810000,33.780000,27.000000,3589.000000,507.000000,1484.000000,495.000000,5.793400,270500.000000
-118.360000,33.820000,28.000000,67.000000,15.000000,49.000000,11.000000,6.135900,330000.000000
-119.670000,36.330000,19.000000,1241.000000,244.000000,850.000000,237.000000,2.937500,81700.000000
-119.560000,36.510000,37.000000,1018.000000,213.000000,663.000000,204.000000,1.663500,67000.000000
-121.430000,38.630000,43.000000,1009.000000,225.000000,604.000000,218.000000,1.664100,67000.000000
-120.650000,35.480000,19.000000,2310.000000,471.000000,1341.000000,441.000000,3.225000,166900.000000
-122.840000,38.400000,15.000000,3080.000000,617.000000,1446.000000,

In [38]:
# Imprimir 50 ultimas lineas de un archivo
!tail -50 sample_data/california_housing_data.csv

tail: cannot open 'sample_data/california_housing_data.csv' for reading: No such file or directory


**Cuál fue el problema?**🤔 corregir y volver a correr.

Que se cambio el nombre del archivo california_housing_test.csv california_housing_data.csv

In [39]:
!tail -50 sample_data/california_housing_test.csv


-122.420000,37.790000,52.000000,3364.000000,1100.000000,2112.000000,1045.000000,2.134300,400000.000000
-122.640000,41.630000,19.000000,2722.000000,479.000000,1108.000000,430.000000,3.106200,100000.000000
-118.020000,33.910000,34.000000,2518.000000,429.000000,1309.000000,421.000000,4.786100,210700.000000
-119.020000,35.360000,48.000000,1833.000000,396.000000,947.000000,363.000000,2.282700,70000.000000
-121.330000,38.650000,23.000000,2446.000000,523.000000,1132.000000,513.000000,2.626600,198500.000000
-118.080000,33.950000,32.000000,1962.000000,387.000000,1274.000000,398.000000,4.830400,160600.000000
-118.080000,33.790000,34.000000,2840.000000,395.000000,1127.000000,396.000000,7.614400,376200.000000
-118.230000,33.910000,27.000000,1694.000000,393.000000,1890.000000,373.000000,3.034100,89100.000000
-118.290000,33.750000,37.000000,1319.000000,292.000000,766.000000,285.000000,2.703100,218900.000000
-118.020000,34.130000,34.000000,1966.000000,319.000000,980.000000,297.000000,7.730700,429000.

In [41]:
# Cuántas lineas tiene el archivo?
!wc -l california_housing_test.csv

wc: california_housing_test.csv: No such file or directory


Aca el error es que estamos dentro del directorio de content/, pero el archivo que queremos contar esta dentro de sample_data, por lo que primero tenemos que entrar al directorio de sample_data y despues al archivo

In [47]:
!pwd
%cd sample_data

/content
/content/sample_data


In [48]:
!wc -l california_housing_test.csv


3001 california_housing_test.csv


In [58]:
# Copiar un archivo.
!cp sample_data/california_housing_train.csv test.csv

cp: cannot stat 'sample_data/california_housing_train.csv': No such file or directory


In [59]:
!pwd
!cp /content/sample_data/california_housing_train.csv /content/test.csv

/content/sample_data


Yo lo mande a la carpeta de content/ en la

In [60]:
# Dónde se guardo?
!pwd
!ls

/content/sample_data
anscombe.json		      mnist_test.csv	     test.csv
california_housing_test.csv   mnist_train_small.csv
california_housing_train.csv  README.md
-rw-r--r-- 1 root root 1706430 Apr  1 15:50 test.csv


In [63]:
# Cambiar el nombre de un archivo
!mv test.csv CopiaCaliforniaTrain.csv

# Ver si el cambio se realizó
!ls

mv: cannot stat 'test.csv': No such file or directory
anscombe.json		      CopiaCaliforniaTrain.csv	README.md
california_housing_test.csv   mnist_test.csv
california_housing_train.csv  mnist_train_small.csv


In [66]:
# Mover el archivo a otra carpeta
!mv CopiaCaliforniaTrain.csv sample_data/CopiaCaliforniaTrain.csv
!mv CopiaCaliforniaTrain.csv /content/sample_data/CopiaCaliforniaTrain.csv


mv: cannot move 'CopiaCaliforniaTrain.csv' to 'sample_data/CopiaCaliforniaTrain.csv': No such file or directory


In [76]:
!ls -l

total 55504
-rwxr-xr-x 1 root root     1697 Jan  1  2000 anscombe.json
-rw-r--r-- 1 root root   301141 Mar 30 13:29 california_housing_test.csv
-rw-r--r-- 1 root root  1706430 Mar 30 13:29 california_housing_train.csv
-rw-r--r-- 1 root root 18289443 Mar 30 13:29 mnist_test.csv
-rw-r--r-- 1 root root 36523880 Mar 30 13:29 mnist_train_small.csv
-rwxr-xr-x 1 root root      962 Jan  1  2000 README.md


In [77]:
!ls -l /content/

total 3340
-rw-r--r-- 1 root root 1706430 Apr  1 15:50 CopiaCaliforniaTrain.csv
drwxr-xr-x 1 root root    4096 Apr  1 16:00 sample_data
-rw-r--r-- 1 root root 1706430 Apr  1 15:55 test.csv


El archivo se movio a la carpeta de content

## 3. Ejercicio 3

En este ejercicio utilizaremos dos herramientas para ejemplificar la manipulación de archivos de texto en Unix, aprender a guardar resultados en archivos y concatenar comandos.


**Awk** es un lenguaje de procesamiento de texto y un comando diseñado para buscar, filtrar y manipular datos en archivos de texto, especialmente tablas o datos separados por delimitadores (como espacios o comas). Permite seleccionar líneas según patrones, extraer columnas, realizar cálculos y generar reportes.

**Grep** es un comando utilizado para buscar líneas que coincidan con un patrón de texto dentro de archivos o flujos de datos. Permite filtrar información rápidamente usando expresiones regulares y opciones como mostrar el número de línea, ignorar mayúsculas/minúsculas, o buscar recursivamente en directorios.


Al igual que muchos programas de Unix, podemos entender los parámetros posibles utilizando un parámetro específico que imprime más información. Para awk este parámetro es **-W help**, pero en la mayoría de los programas es **--help**. Para más información, recurrir a guías en internet, por [ejemplo](https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://www.geeksforgeeks.org/linux-unix/awk-command-unixlinux-examples/&ved=2ahUKEwiC_MzmioiQAxWeBLkGHS51Dh8QFnoECBkQAQ&usg=AOvVaw3tvNrjj9AJqSclSDOodvpg).


In [79]:
!pwd

/content/sample_data


In [81]:
%cd /content/sample_data/

# Me interesan únicamente las columnas "longitude" y "latitude" que están en la columna 1 y 2 del archivo california_housing_test
# Ver primeras lineas
!head -2 california_housing_test.csv


/content/sample_data
"longitude","latitude","housing_median_age","total_rooms","total_bedrooms","population","households","median_income","median_house_value"
-122.050000,37.370000,27.000000,3885.000000,661.000000,1537.000000,606.000000,6.608500,344700.000000


In [82]:
# Imprimir únicamente las columnas de interés
!awk -F"," '{print $1,$2}' california_housing_test.csv

"longitude" "latitude"
-122.050000 37.370000
-118.300000 34.260000
-117.810000 33.780000
-118.360000 33.820000
-119.670000 36.330000
-119.560000 36.510000
-121.430000 38.630000
-120.650000 35.480000
-122.840000 38.400000
-118.020000 34.080000
-118.240000 33.980000
-119.120000 35.850000
-121.930000 37.250000
-117.030000 32.970000
-117.970000 33.730000
-117.990000 33.810000
-120.810000 37.530000
-121.200000 38.690000
-118.880000 34.210000
-122.590000 38.010000
-122.150000 37.750000
-121.370000 38.680000
-118.160000 34.070000
-122.200000 37.790000
-117.280000 33.280000
-118.030000 34.160000
-122.420000 37.760000
-118.390000 33.990000
-118.450000 34.070000
-118.480000 34.010000
-119.350000 36.330000
-118.300000 33.910000
-121.130000 39.310000
-118.080000 34.550000
-118.320000 33.940000
-118.110000 34.000000
-122.530000 37.970000
-118.020000 33.920000
-118.050000 33.930000
-119.010000 34.230000
-119.320000 36.200000
-116.920000 32.770000
-118.060000 34.150000
-117.270000 34.090000
-118.2300

Desglocemos este comando:  

* **-F ","** *indica que las columnas se encuentran separadas por coma.*
* **'{print .... '**  *indica al programa a imprimir el resultado.*
* **\$1, \$2**    *indica al programa imprimir únicamente las columnas 1 y 2.*


In [83]:
!awk -W help

Usage: mawk [Options] [Program] [file ...]

Program:
    The -f option value is the name of a file containing program text.
    If no -f option is given, a "--" ends option processing; the following
    parameters are the program text.

Options:
    -f program-file  Program  text is read from file instead of from the
                     command-line.  Multiple -f options are accepted.
    -F value         sets the field separator, FS, to value.
    -v var=value     assigns value to program variable var.
    --               unambiguous end of options.

    Implementation-specific options are prefixed with "-W".  They can be
    abbreviated:

    -W version       show version information and exit.
    -W dump          show assembler-like listing of program and exit.
    -W help          show this message and exit.
    -W interactive   set unbuffered output, line-buffered input.
    -W exec file     use file as program as well as last option.
    -W random=number set initial random seed

Para el próximo ejercicio, sólo nos interesa el dato de la población en las coordenadas "-118.140000,33.880000", para encontrarlo entre las 3001 líneas posibles, utilizaremos el comando grep.

In [84]:
# Encontrar líneas que presentan las coordenadas de interés

!grep -- "-118.140000,33.880000" california_housing_test.csv

-118.140000,33.880000,41.000000,1531.000000,343.000000,1119.000000,341.000000,4.364600,161400.000000
-118.140000,33.880000,30.000000,2596.000000,580.000000,1662.000000,539.000000,4.050700,179500.000000


In [85]:
# Aislar el dato de Población
!grep -- "-118.140000,33.880000" california_housing_test.csv | awk -F "," '{print $6}'

1119.000000
1662.000000


En este caso utilizamos dos comandos simultaneamente.
Grep para buscar la información de interés, y awk para imprimir únicamente la columna "Población". El **símbolo |** significa que el resultado de un programa es enviado al siguiente. De esta manera, muchos análisis pueden ser concatenados utilizando este símbolo.

A continuación guardaremos el resultado en un archivo llamado poblacion.txt, con el **símbolo >**.

In [86]:
# Guardar el resultado en un archivo
!grep -- "-118.140000,33.880000" california_housing_test.csv | awk -F "," '{print $6}' > poblacion.txt

In [87]:
# Se creó el archivo? qué tiene adentro?
!ls





anscombe.json		      mnist_test.csv	     README.md
california_housing_test.csv   mnist_train_small.csv
california_housing_train.csv  poblacion.txt


Se creo el archivo poblacion.txt

In [88]:
!cat poblacion.txt

1119.000000
1662.000000


Contiene la informacion de la columna 6, la poblacion de las filas que tenian las cordenadas de longitud y latitud "118.140000,33.880000"

## 4. Ejercicio 4

Un ambiente virtual es un espacio aislado donde puedes instalar paquetes y dependencias sin afectar el sistema principal ni otros proyectos. Esto permite trabajar con distintas versiones de bibliotecas y programas en paralelo, evitando conflictos. Existen varios programas que crean ambientes virtuales, entre ellos conda, mamba, micromamba.

A lo largo del curso utilizaremos diferentes ambientes virtuales de **micromamba**. Variantes de los siguientes comandos estarán presentes en la mayoría de los trabajos prácticos.

In [89]:
# Para el pŕoximo ejercicio, necesitamos el programa seqkit
!seqkit

/bin/bash: line 1: seqkit: command not found


**Qué significa este error?** 🤔

Que no tenemos el programa instalado

In [94]:
# Lo instalaremos con micromamba
# En primer lugar, descargar micromamba
!curl -Ls https://micro.mamba.pm/api/micromamba/linux-64/latest | tar -xvj bin/micromamba

# Agregar micromamba al ambiente
import os
os.environ['PATH'] = './bin/:' + os.environ['PATH']

 #para ver si funciono
!echo $PATH

bin/micromamba
./bin/:./bin/:./bin/:/opt/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/tools/node/bin:/tools/google-cloud-sdk/bin


In [95]:
# Crear entorno llamado "colab" e instalar el programa seqkit
!./bin/micromamba create -y -n colab -c bioconda seqkit


[+] 0.0s
[+] 0.0s
[+] 0.1s
bioconda/linux-64   1%
bioconda/noarch     7%bioconda/noarch                                   
bioconda/linux-64                                 


Transaction

  Prefix: /root/.local/share/mamba/envs/colab

  Updating specs:

   - seqkit


  Package   Version  Build       Channel      Size
────────────────────────────────────────────────────
  Install:
────────────────────────────────────────────────────

  + seqkit   2.13.0  he881be0_0  bioconda      7MB

  Summary:

  Install: 1 packages

  Total download: 7MB

────────────────────────────────────────────────────



Transaction starting
[+] 0.0s
Extracting         0%seqkit                                               6.9MB @   1.3MB/s  0.0s
[+] 0.1s
Extracting   (1)  ⣾  Linking seqkit-2.13.0-he881be0_0

Transaction finished


To activate this environment, use:

    micromamba activate colab

Or to execute a single command in this environment, use:

    micromamba run -n colab mycommand



In [96]:
!seqkit

/bin/bash: line 1: seqkit: command not found


**Por qué no funciona?** 🤔

Para llamar un programa instalado con conda debemos utilizar el comando **!micromamba run** y el nombre del ambiente, antes del comando del programa de interés.

In [97]:
!micromamba run -n colab seqkit

SeqKit -- a cross-platform and ultrafast toolkit for FASTA/Q file manipulation

Version: 2.13.0

Author: Wei Shen <shenwei356@gmail.com>

Documents  : http://bioinf.shenwei.me/seqkit
Source code: https://github.com/shenwei356/seqkit
Please cite: https://doi.org/10.1002/imt2.191


Seqkit utilizes the pgzip (https://github.com/klauspost/pgzip) package to
read and write gzip file, and the outputted gzip file would be slighty
larger than files generated by GNU gzip.

Seqkit writes gzip files very fast, much faster than the multi-threaded pigz,
therefore there's no need to pipe the result to gzip/pigz.

Seqkit also supports reading and writing xz (.xz) and zstd (.zst) formats since v2.2.0.
Bzip2 format is supported since v2.4.0.

Compression level:
  format   range   default  comment
  gzip     1-9     5        https://github.com/klauspost/pgzip sets 5 as the default value.
  xz       NA      NA       https://github.com/ulikunitz/xz does not support.
  zstd     1-4     2        roughly equa

In [98]:
# Cree un ambiente llamado test con el programa seqtk
!./bin/micromamba create -y -n test -c bioconda seqkit


bioconda/linux-64                                           Using cache
bioconda/noarch                                             Using cache


Transaction

  Prefix: /root/.local/share/mamba/envs/test

  Updating specs:

   - seqkit


  Package   Version  Build       Channel        Size
──────────────────────────────────────────────────────
  Install:
──────────────────────────────────────────────────────

  + seqkit   2.13.0  he881be0_0  bioconda     Cached

  Summary:

  Install: 1 packages

  Total download: 0 B

──────────────────────────────────────────────────────



Transaction starting
[+] 0.0s
Linking seqkit-2.13.0-he881be0_0

Transaction finished


To activate this environment, use:

    micromamba activate test

Or to execute a single command in this environment, use:

    micromamba run -n test mycommand



In [103]:
!./bin/micromamba run -n test seqkit


SeqKit -- a cross-platform and ultrafast toolkit for FASTA/Q file manipulation

Version: 2.13.0

Author: Wei Shen <shenwei356@gmail.com>

Documents  : http://bioinf.shenwei.me/seqkit
Source code: https://github.com/shenwei356/seqkit
Please cite: https://doi.org/10.1002/imt2.191


Seqkit utilizes the pgzip (https://github.com/klauspost/pgzip) package to
read and write gzip file, and the outputted gzip file would be slighty
larger than files generated by GNU gzip.

Seqkit writes gzip files very fast, much faster than the multi-threaded pigz,
therefore there's no need to pipe the result to gzip/pigz.

Seqkit also supports reading and writing xz (.xz) and zstd (.zst) formats since v2.2.0.
Bzip2 format is supported since v2.4.0.

Compression level:
  format   range   default  comment
  gzip     1-9     5        https://github.com/klauspost/pgzip sets 5 as the default value.
  xz       NA      NA       https://github.com/ulikunitz/xz does not support.
  zstd     1-4     2        roughly equa

## 5. Ejercicio extra

Google colab permite ejecutar código en diferentes lenguajes siempre y cuando esté bien especificado y escrito en la sintaxis correspondiente. El lenguaje por defecto es Python.

A continuación veremos como ejecutar diferentes lenguajes.

In [99]:
# instalar R
!pip install rpy2
%load_ext rpy2.ipython

In [100]:
# Ejecutar un código simple en R
%%R

suma <- 2 + 3
print("El resultado de la suma es:")
print(suma)

[1] "El resultado de la suma es:"
[1] 5


In [101]:
# Ejecutar un código simple en Python

suma = 2 + 3
print("El resultado de la suma es:")
print(suma)


El resultado de la suma es:
5


In [102]:
# Ejecutar un código simple en Perl
%%perl

my $suma = 2 + 3;
print "El resultado de la suma es:\n";
print "$suma\n";

El resultado de la suma es:
5


**Qué diferencias logra percibir?** 🤔



Se pueden observar varias diferencias entre los códigos ejecutados en R, Python y Perl, aunque todos realizan la misma operación básica.

En primer lugar, en R el resultado se muestra precedido por [1], lo cual indica la posición del elemento dentro de un vector. Esto refleja que R trabaja de manera nativa con estructuras vectoriales.

En Python, en cambio, la salida es más simple y directa, sin ningún indicador adicional, ya que imprime exactamente lo que se le indica con la función print().

Por otro lado, en Perl, aunque el resultado final es el mismo, se observa una sintaxis distinta, como el uso del símbolo $ para declarar variables y la inclusión explícita de \n para generar saltos de línea.

Además, otra diferencia importante es el uso de los llamados magic commands en entornos como Google Colab, donde se utilizan %%R y %%perl para ejecutar código en distintos lenguajes dentro del mismo notebook, mientras que Python es el lenguaje nativo y no requiere un prefijo especial.

En Google Colab es posible mezclar distintos lenguajes de programación dentro de un mismo cuaderno. Aunque las celdas se ejecutan por defecto en Python, se pueden usar comandos especiales (magics) para ejecutar otros lenguajes como R (%%R), Bash (! o %%bash), JavaScript (%%javascript), entre otros. Esto permite integrar análisis, automatización y visualización en un solo entorno interactivo.  

Los objetos/variables de python pueden ser llamadas dentro de otros lenguajes, utilizando el símbolo $ o {}

**El siguiente código combina python y bash, lo podes distinguir?** 🤔


In [104]:
basedir = "./"
res = "resultados"

dir = basedir + res

!mkdir $dir

!ls {basedir}

anscombe.json		     california_housing_train.csv  poblacion.txt
bin			     mnist_test.csv		   README.md
california_housing_test.csv  mnist_train_small.csv	   resultados


El lenguaje de bash utiliza el simbolo ! para indicar la salida de la terminal, mientras que phyton no lo utiliza
